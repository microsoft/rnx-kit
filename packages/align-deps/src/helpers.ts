import type { PackageManifest } from "@rnx-kit/tools-node/package";
import { writePackage } from "@rnx-kit/tools-node/package";
import detectIndent from "detect-indent";
import fs from "fs";
import semverValidRange from "semver/ranges/valid";

export function compare<T>(lhs: T, rhs: T): -1 | 0 | 1 {
  if (lhs === rhs) {
    return 0;
  } else if (lhs < rhs) {
    return -1;
  } else {
    return 1;
  }
}

export function dropPatchFromVersion(version: string): string {
  return version
    .split("||")
    .map((input) => {
      const versionRange = input.trim();
      if (!semverValidRange(versionRange)) {
        throw new Error(`Invalid version number: ${versionRange}`);
      }

      if (!versionRange) {
        return "*";
      }

      return versionRange
        .split(" ")
        .map((v) => {
          if (v === "*" || v === "-") {
            // No need to manipulate `*` or hyphen ranges, e.g. `1.0 - 2.0`
            return v;
          }

          const [major, minor = "0"] = v.split(".");
          return major === "^0" || major === "~0"
            ? `0.${minor}`
            : `${major}.${minor}`;
        })
        .join(" ");
    })
    .join(" || ");
}

export function modifyManifest(
  pkgPath: string,
  manifest: PackageManifest
): void {
  const content = fs.readFileSync(pkgPath, { encoding: "utf-8" });
  const indent = detectIndent(content).indent || "  ";
  writePackage(pkgPath, manifest, indent);
}

export function omitEmptySections(manifest: PackageManifest): PackageManifest {
  const sections = ["dependencies", "peerDependencies", "devDependencies"];
  for (const sectionName of sections) {
    const section = manifest[sectionName];
    if (typeof section === "object" && Object.keys(section).length === 0) {
      delete manifest[sectionName];
    }
  }
  return manifest;
}
