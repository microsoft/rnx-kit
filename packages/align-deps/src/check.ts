import type { KitConfig } from "@rnx-kit/config";
import { getKitCapabilities, getKitConfig } from "@rnx-kit/config";
import { error, info, warn } from "@rnx-kit/console";
import { isPackageManifest, readPackage } from "@rnx-kit/tools-node/package";
import chalk from "chalk";
import { diffLinesUnified } from "jest-diff";
import path from "path";
import { migrateConfig } from "./compatibility/config";
import { gatherRequirements, getRequirements } from "./dependencies";
import { findBadPackages } from "./findBadPackages";
import { modifyManifest } from "./helpers";
import { updatePackageManifest } from "./manifest";
import { filterPreset, mergePresets } from "./preset";
import { resolveCustomProfiles } from "./profiles";
import type {
  AlignDepsConfig,
  CheckConfig,
  CheckOptions,
  Command,
  ErrorCode,
  Options,
  Preset,
} from "./types";

export function containsValidPresets(config: KitConfig["alignDeps"]): boolean {
  const presets = config?.presets;
  return !presets || (Array.isArray(presets) && presets.length > 0);
}

export function containsValidRequirements(
  config: KitConfig["alignDeps"]
): boolean {
  const requirements = config?.requirements;
  if (requirements) {
    if (Array.isArray(requirements)) {
      return requirements.length > 0;
    } else if (typeof requirements === "object") {
      return (
        Array.isArray(requirements.production) &&
        requirements.production.length > 0
      );
    }
  }
  return false;
}

function ensurePreset(preset: Preset, requirements: string[]): void {
  if (Object.keys(preset).length === 0) {
    throw new Error(
      `No profiles could satisfy requirements: ${requirements.join(", ")}`
    );
  }
}

export function getCheckConfig(
  manifestPath: string,
  {
    loose,
    uncheckedReturnCode = 0,
    supportedVersions,
    targetVersion,
  }: CheckOptions
): number | CheckConfig {
  const manifest = readPackage(manifestPath);
  if (!isPackageManifest(manifest)) {
    error(
      `'${manifestPath}' does not contain a valid package manifest - please make sure it's not missing 'name' or 'version'`
    );
    return 1;
  }

  const badPackages = findBadPackages(manifest);
  if (badPackages) {
    warn(
      `Known bad packages are found in '${manifest.name}':\n` +
        badPackages
          .map((pkg) => `    ${pkg.name}@${pkg.version}: ${pkg.reason}`)
          .join("\n")
    );
  }

  const projectRoot = path.dirname(manifestPath);
  const kitConfig = getKitConfig({ cwd: projectRoot });
  if (!kitConfig) {
    return uncheckedReturnCode;
  }

  const {
    reactNativeVersion: targetReactNativeVersion,
    reactNativeDevVersion,
    kitType,
    capabilities: targetCapabilities,
    customProfiles,
  } = getKitCapabilities({
    // React Native versions declared in the package's config should always
    // override the ones specified with the `--vigilant` flag.
    ...(supportedVersions
      ? { reactNativeVersion: supportedVersions }
      : undefined),
    // We should not set dev version if the package is configured. It may have
    // been intentionally left out to reuse `reactNativeVersion`.
    ...(!kitConfig.reactNativeVersion && targetVersion
      ? { reactNativeDevVersion: targetVersion }
      : undefined),
    ...kitConfig,
  });

  const customProfilesPath = resolveCustomProfiles(projectRoot, customProfiles);

  const { reactNativeVersion, capabilities: requiredCapabilities } =
    getRequirements(
      targetReactNativeVersion,
      kitType,
      manifest,
      projectRoot,
      customProfilesPath,
      { loose }
    );
  requiredCapabilities.push(...targetCapabilities);

  return {
    kitType,
    reactNativeVersion,
    reactNativeDevVersion,
    capabilities: requiredCapabilities,
    customProfilesPath,
    manifest,
  };
}

export function getConfig(
  manifestPath: string
): AlignDepsConfig | CheckConfig | ErrorCode {
  const manifest = readPackage(manifestPath);
  if (!isPackageManifest(manifest)) {
    return "invalid-manifest";
  }

  const badPackages = findBadPackages(manifest);
  if (badPackages) {
    warn(
      `Known bad packages are found in '${manifest.name}':\n` +
        badPackages
          .map((pkg) => `\t${pkg.name}@${pkg.version}: ${pkg.reason}`)
          .join("\n")
    );
  }

  const projectRoot = path.dirname(manifestPath);
  const kitConfig = getKitConfig({ cwd: projectRoot });
  if (!kitConfig) {
    return "not-configured";
  }

  const { kitType = "library", alignDeps, ...config } = kitConfig;
  if (alignDeps) {
    const errors = [];
    if (!containsValidPresets(alignDeps)) {
      errors.push(`${manifestPath}: 'alignDeps.presets' cannot be empty`);
    }
    if (!containsValidRequirements(alignDeps)) {
      errors.push(`${manifestPath}: 'alignDeps.requirements' cannot be empty`);
    }
    if (errors.length > 0) {
      for (const e of errors) {
        error(e);
      }
      return "invalid-configuration";
    }
    return {
      kitType,
      alignDeps: {
        presets: ["microsoft/react-native"],
        requirements: [],
        capabilities: [],
        ...alignDeps,
      },
      ...config,
      manifest,
    };
  }

  const {
    capabilities,
    customProfiles,
    reactNativeDevVersion,
    reactNativeVersion,
  } = getKitCapabilities(config);

  return {
    kitType,
    reactNativeVersion,
    ...(config.reactNativeDevVersion ? { reactNativeDevVersion } : undefined),
    capabilities,
    customProfiles,
    manifest,
  } as CheckConfig;
}

function resolve(
  { kitType, alignDeps, manifest }: AlignDepsConfig,
  projectRoot: string,
  options: CheckOptions
) {
  const { capabilities, presets, requirements } = alignDeps;

  const prodRequirements = Array.isArray(requirements)
    ? requirements
    : requirements.production;
  const mergedPreset = mergePresets(presets, projectRoot);
  const initialProdPreset = filterPreset(prodRequirements, mergedPreset);
  ensurePreset(initialProdPreset, prodRequirements);

  const devPreset = (() => {
    if (kitType === "app") {
      // Preset for development is unused when the package is an app.
      return {};
    } else if (Array.isArray(requirements)) {
      return initialProdPreset;
    } else {
      const devRequirements = requirements.development;
      const devPreset = filterPreset(devRequirements, mergedPreset);
      ensurePreset(devPreset, devRequirements);
      return devPreset;
    }
  })();

  if (kitType === "app") {
    const { preset: prodMergedPreset, capabilities: mergedCapabilities } =
      gatherRequirements(
        projectRoot,
        manifest,
        initialProdPreset,
        capabilities,
        options
      );
    return {
      devPreset,
      prodPreset: prodMergedPreset,
      capabilities: mergedCapabilities,
    };
  }

  return { devPreset, prodPreset: initialProdPreset, capabilities };
}

export function checkPackageManifest(
  manifestPath: string,
  options: CheckOptions,
  inputConfig = getConfig(manifestPath)
): ErrorCode {
  if (typeof inputConfig === "string") {
    return inputConfig;
  }

  const config = migrateConfig(inputConfig);
  const { devPreset, prodPreset, capabilities } = resolve(
    config,
    path.dirname(manifestPath),
    options
  );
  if (capabilities.length === 0) {
    return "success";
  }

  const { kitType, manifest } = config;

  if (kitType === "app") {
    info(
      "Aligning dependencies according to the following profiles:",
      Object.keys(prodPreset).join(", ")
    );
  } else {
    info("Aligning dependencies according to the following profiles:");
    info("\t- Development:", Object.keys(devPreset).join(", "));
    info("\t- Production:", Object.keys(prodPreset).join(", "));
  }

  const updatedManifest = updatePackageManifest(
    manifest,
    capabilities,
    Object.values(prodPreset),
    Object.values(devPreset),
    kitType
  );

  // Don't fail when manifests only have whitespace differences.
  const updatedManifestJson = JSON.stringify(updatedManifest, undefined, 2);
  const normalizedManifestJson = JSON.stringify(manifest, undefined, 2);

  if (updatedManifestJson !== normalizedManifestJson) {
    if (options.write) {
      modifyManifest(manifestPath, updatedManifest);
    } else {
      const diff = diffLinesUnified(
        normalizedManifestJson.split("\n"),
        updatedManifestJson.split("\n"),
        {
          aAnnotation: "Current",
          aColor: chalk.red,
          bAnnotation: "Expected",
          bColor: chalk.green,
        }
      );
      console.log(diff);
      return "unsatisfied";
    }
  }

  return "success";
}

export function makeCheckCommand(options: Options): Command {
  return (manifest: string) => checkPackageManifest(manifest, options);
}
