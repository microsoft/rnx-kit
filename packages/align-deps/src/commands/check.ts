import { info } from "@rnx-kit/console";
import { pickValues } from "@rnx-kit/tools-language";
import type { PackageManifest } from "@rnx-kit/tools-node/package";
import { readPackage } from "@rnx-kit/tools-node/package";
import chalk from "chalk";
import { diffLinesUnified } from "jest-diff";
import * as path from "path";
import { migrateConfig } from "../compatibility/config";
import { loadConfig } from "../config";
import { isError } from "../errors";
import { modifyManifest } from "../helpers";
import { updatePackageManifest } from "../manifest";
import { resolve } from "../preset";
import type { Command, ErrorCode, Options } from "../types";
import { checkPackageManifestUnconfigured } from "./vigilant";

const visibleKeys = [
  "name",
  "version",
  "dependencies",
  "peerDependencies",
  "devDependencies",
];

function stringify(manifest: PackageManifest): string {
  return JSON.stringify(pickValues(manifest, visibleKeys), undefined, 2);
}

/**
 * Checks the specified package manifest for misaligned dependencies.
 *
 * There are essentially two modes of operation depending on whether the package
 * is an app or a library.
 *
 * - For libraries, only dependencies that are declared under capabilities are
 *   checked. `align-deps` will ensure that `peerDependencies` and
 *   `devDependencies` are correctly used to satisfy the declared capabilities.
 * - For apps, its dependencies and the dependencies of its dependencies are
 *   checked. `align-deps` will ensure that `dependencies` and `devDependencies`
 *   are correctly used to satisfy the declared capabilities. Additionally,
 *   requirements may only resolve to a single profile. If multiple profiles
 *   satisfy the requirements, the command will fail.
 *
 * @see {@link updatePackageManifest}
 *
 * @param manifestPath Path to the package manifest to check
 * @param options Command line options
 * @param inputConfig Configuration in the package manifest
 * @returns `success` when everything is in order; an {@link ErrorCode} otherwise
 */
export function checkPackageManifest(
  manifestPath: string,
  options: Options,
  inputConfig = loadConfig(manifestPath)
): ErrorCode {
  if (isError(inputConfig)) {
    return inputConfig;
  }

  const config = migrateConfig(inputConfig, manifestPath, options);
  const { devPreset, prodPreset, capabilities } = resolve(
    config,
    path.dirname(manifestPath),
    options
  );
  const { kitType, manifest } = config;

  if (kitType === "app" && Object.keys(prodPreset).length !== 1) {
    return "invalid-app-requirements";
  } else if (capabilities.length === 0) {
    return "success";
  }

  if (options.verbose) {
    if (kitType === "app") {
      info(
        `${manifestPath}: Aligning your app's dependencies according to the following profiles:`,
        Object.keys(prodPreset).join(", ")
      );
    } else {
      info(
        `${manifestPath}: Aligning your library's dependencies according to the following profiles:\n` +
          `\t- Development: ${Object.keys(devPreset).join(", ")}\n` +
          `\t- Production: ${Object.keys(prodPreset).join(", ")}`
      );
    }
  }

  const updatedManifest = updatePackageManifest(
    manifestPath,
    manifest,
    capabilities,
    prodPreset,
    devPreset,
    kitType
  );

  // Don't fail when manifests only have whitespace differences.
  const updatedManifestJson = stringify(updatedManifest);
  const normalizedManifestJson = stringify(manifest);

  if (updatedManifestJson !== normalizedManifestJson) {
    if (options.write) {
      modifyManifest(manifestPath, updatedManifest);
    } else {
      const diff = diffLinesUnified(
        normalizedManifestJson.split("\n"),
        updatedManifestJson.split("\n"),
        {
          aAnnotation: "Current",
          aColor: chalk.red,
          bAnnotation: "Expected",
          bColor: chalk.green,
        }
      );
      console.log(diff);
      return "unsatisfied";
    }
  }

  return "success";
}

/**
 * Creates the check command. This is the default command no other flags are
 * specified.
 *
 * In normal mode, `align-deps` will only check packages that have a
 * configuration, and only listed capabilities.
 *
 * In vigilant mode, `align-deps` will check all packages in the workspace,
 * regardless of whether they have a configuration. For packages that do have a
 * configuration, the listed capabilities will be checked first as usual. The
 * remaining capabilities will then be checked, but are treated as unconfigured.
 *
 * @see {@link checkPackageManifest}
 * @see {@link checkPackageManifestUnconfigured}
 *
 * @param options Command line options
 * @returns The check command
 */
export function makeCheckCommand(options: Options): Command {
  const { presets, requirements } = options;
  if (!requirements) {
    return (manifest: string) => checkPackageManifest(manifest, options);
  }

  return (manifest: string) => {
    const inputConfig = loadConfig(manifest);
    const config = isError(inputConfig)
      ? inputConfig
      : migrateConfig(inputConfig, manifest, options);

    // If the package is configured, run the normal check first.
    if (!isError(config)) {
      const result = checkPackageManifest(manifest, options, config);
      return result !== "success"
        ? result
        : checkPackageManifestUnconfigured(manifest, options, config);
    }

    // Otherwise, run the unconfigured check only.
    if (config === "invalid-configuration" || config === "not-configured") {
      // In "vigilant" mode, we allow packages to declare which presets should
      // be used in config, overriding the `--presets` flag.
      return checkPackageManifestUnconfigured(manifest, options, {
        kitType: "library",
        alignDeps: {
          presets,
          requirements,
          capabilities: [],
        },
        manifest: readPackage(manifest),
      });
    }

    return config;
  };
}
