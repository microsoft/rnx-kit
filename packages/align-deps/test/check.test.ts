import { checkPackageManifest } from "../src/commands/check";
import { defaultConfig } from "../src/config";
import profile_0_68 from "../src/presets/microsoft/react-native/profile-0.68";
import profile_0_69 from "../src/presets/microsoft/react-native/profile-0.69";
import profile_0_70 from "../src/presets/microsoft/react-native/profile-0.70";
import { packageVersion } from "./helpers";

jest.mock("fs");

const defaultOptions = {
  presets: defaultConfig.presets,
  loose: false,
  migrateConfig: false,
  verbose: false,
  write: false,
};

const writeOptions = {
  ...defaultOptions,
  write: true,
};

describe("checkPackageManifest({ kitType: 'library' })", () => {
  const rnxKitConfig = require("@rnx-kit/config");
  const fs = require("fs");

  const consoleLogSpy = jest.spyOn(global.console, "log");
  const consoleWarnSpy = jest.spyOn(global.console, "warn");

  const mockManifest = {
    name: "@rnx-kit/align-deps",
    version: "0.0.1",
  };

  const react_v68_v69_v70 = [
    packageVersion(profile_0_68, "react"),
    packageVersion(profile_0_69, "react"),
    packageVersion(profile_0_70, "react"),
  ].join(" || ");

  const v68_v69_v70 = [
    packageVersion(profile_0_68, "core"),
    packageVersion(profile_0_69, "core"),
    packageVersion(profile_0_70, "core"),
  ].join(" || ");

  beforeEach(() => {
    fs.__setMockContent({});
    rnxKitConfig.__setMockConfig();
    consoleLogSpy.mockReset();
    consoleWarnSpy.mockReset();
  });

  afterAll(() => {
    jest.clearAllMocks();
  });

  test("returns error code when reading invalid manifests", () => {
    const result = checkPackageManifest("package.json", defaultOptions);
    expect(result).toBe("invalid-manifest");
    expect(consoleLogSpy).not.toBeCalled();
    expect(consoleWarnSpy).not.toBeCalled();
  });

  test("returns early if 'rnx-kit' is missing from the manifest", () => {
    fs.__setMockContent({
      ...mockManifest,
      dependencies: { "react-native-linear-gradient": "0.0.0" },
    });

    const result = checkPackageManifest("package.json", defaultOptions);

    expect(result).toBe("not-configured");
    expect(consoleLogSpy).not.toBeCalled();
    expect(consoleWarnSpy).toBeCalled();
  });

  test("prints warnings when detecting bad packages", () => {
    fs.__setMockContent({
      ...mockManifest,
      dependencies: { "react-native-linear-gradient": "0.0.0" },
      peerDependencies: {
        "react-native": profile_0_70["core"],
      },
      devDependencies: {
        "react-native": profile_0_70["core"],
      },
    });
    rnxKitConfig.__setMockConfig({
      alignDeps: { requirements: ["react-native@0.70"] },
    });

    const result = checkPackageManifest("package.json", defaultOptions);

    expect(result).toBe("success");
    expect(consoleLogSpy).not.toBeCalled();
    expect(consoleWarnSpy).toBeCalled();
  });

  test("prints warnings when detecting bad packages (with version range)", () => {
    fs.__setMockContent({
      ...mockManifest,
      dependencies: { "react-native-linear-gradient": "0.0.0" },
    });
    rnxKitConfig.__setMockConfig({
      alignDeps: { requirements: ["react-native@^0.69.0 || ^0.70.0"] },
    });

    const result = checkPackageManifest("package.json", defaultOptions);

    expect(result).toBe("success");
    expect(consoleLogSpy).not.toBeCalled();
    expect(consoleWarnSpy).toBeCalled();
  });

  test("returns early if no capabilities are defined", () => {
    fs.__setMockContent(mockManifest);
    rnxKitConfig.__setMockConfig({
      alignDeps: { requirements: ["react-native@0.70"] },
    });

    const result = checkPackageManifest("package.json", defaultOptions);

    expect(result).toBe("success");
    expect(consoleLogSpy).not.toBeCalled();
    expect(consoleWarnSpy).not.toBeCalled();
  });

  test("returns if no changes are needed", () => {
    fs.__setMockContent({
      ...mockManifest,
      peerDependencies: {
        react: packageVersion(profile_0_70, "react"),
        "react-native": packageVersion(profile_0_70, "core"),
      },
      devDependencies: {
        react: packageVersion(profile_0_70, "react"),
        "react-native": packageVersion(profile_0_70, "core"),
      },
    });
    rnxKitConfig.__setMockConfig({
      alignDeps: {
        requirements: ["react-native@0.70"],
        capabilities: ["core-ios"],
      },
    });

    const result = checkPackageManifest("package.json", defaultOptions);

    expect(result).toBe("success");
    expect(consoleLogSpy).not.toBeCalled();
    expect(consoleWarnSpy).not.toBeCalled();
  });

  test("prints additional information with `--verbose`", () => {
    fs.__setMockContent({
      ...mockManifest,
      peerDependencies: {
        react: packageVersion(profile_0_70, "react"),
        "react-native": packageVersion(profile_0_70, "core"),
      },
      devDependencies: {
        react: packageVersion(profile_0_70, "react"),
        "react-native": packageVersion(profile_0_70, "core"),
      },
    });
    rnxKitConfig.__setMockConfig({
      alignDeps: {
        requirements: ["react-native@0.70"],
        capabilities: ["core-ios"],
      },
    });

    const result = checkPackageManifest("package.json", {
      ...defaultOptions,
      verbose: true,
    });

    expect(result).toBe("success");
    expect(consoleLogSpy).toBeCalled();
    expect(consoleWarnSpy).not.toBeCalled();
  });

  test("returns if no changes are needed (write: true)", () => {
    let didWriteToPath = false;

    fs.__setMockContent({
      ...mockManifest,
      peerDependencies: {
        react: packageVersion(profile_0_70, "react"),
        "react-native": packageVersion(profile_0_70, "core"),
      },
      devDependencies: {
        react: packageVersion(profile_0_70, "react"),
        "react-native": packageVersion(profile_0_70, "core"),
      },
    });
    fs.__setMockFileWriter((p, _content) => {
      didWriteToPath = p;
    });
    rnxKitConfig.__setMockConfig({
      alignDeps: {
        requirements: ["react-native@0.70"],
        capabilities: ["core-ios"],
      },
    });

    expect(checkPackageManifest("package.json", writeOptions)).toBe("success");
    expect(didWriteToPath).toBe(false);
    expect(consoleLogSpy).not.toBeCalled();
    expect(consoleWarnSpy).not.toBeCalled();
  });

  test("returns error code if changes are needed", () => {
    fs.__setMockContent(mockManifest);
    rnxKitConfig.__setMockConfig({
      alignDeps: {
        requirements: ["react-native@0.70"],
        capabilities: ["core-ios"],
      },
    });

    const result = checkPackageManifest("package.json", defaultOptions);

    expect(result).not.toBe("success");
    expect(consoleLogSpy).toBeCalled();
    expect(consoleWarnSpy).not.toBeCalled();
  });

  test("writes changes back to 'package.json'", () => {
    let didWriteToPath = false;

    fs.__setMockContent(mockManifest);
    fs.__setMockFileWriter((p, _content) => {
      didWriteToPath = p;
    });
    rnxKitConfig.__setMockConfig({
      alignDeps: {
        requirements: ["react-native@0.70"],
        capabilities: ["core-ios"],
      },
    });

    expect(checkPackageManifest("package.json", writeOptions)).toBe("success");
    expect(didWriteToPath).toBe("package.json");
    expect(consoleLogSpy).not.toBeCalled();
    expect(consoleWarnSpy).not.toBeCalled();
  });

  test("preserves indentation in 'package.json'", () => {
    let output = "";

    fs.__setMockContent(mockManifest, "\t");
    fs.__setMockFileWriter((_, content) => {
      output = content;
    });
    rnxKitConfig.__setMockConfig({
      alignDeps: {
        requirements: ["react-native@0.70"],
        capabilities: ["core-ios"],
      },
    });

    expect(checkPackageManifest("package.json", writeOptions)).toBe("success");
    expect(output).toMatchSnapshot();
    expect(consoleLogSpy).not.toBeCalled();
    expect(consoleWarnSpy).not.toBeCalled();
  });

  test("uses minimum supported version as development version", () => {
    fs.__setMockContent({
      ...mockManifest,
      peerDependencies: {
        react: react_v68_v69_v70,
        "react-native": v68_v69_v70,
      },
      devDependencies: {
        react: packageVersion(profile_0_68, "react"),
        "react-native": packageVersion(profile_0_68, "core"),
      },
    });
    rnxKitConfig.__setMockConfig({
      alignDeps: {
        requirements: ["react-native@0.68 || 0.69 || 0.70"],
        capabilities: ["core-ios"],
      },
    });

    const result = checkPackageManifest("package.json", defaultOptions);

    expect(result).toBe("success");
    expect(consoleLogSpy).not.toBeCalled();
    expect(consoleWarnSpy).not.toBeCalled();
  });

  test("uses declared development version", () => {
    fs.__setMockContent({
      ...mockManifest,
      peerDependencies: {
        react: react_v68_v69_v70,
        "react-native": v68_v69_v70,
      },
      devDependencies: {
        react: packageVersion(profile_0_69, "react"),
        "react-native": packageVersion(profile_0_69, "core"),
      },
    });
    rnxKitConfig.__setMockConfig({
      alignDeps: {
        requirements: {
          development: ["react-native@0.69"],
          production: ["react-native@0.68 || 0.69 || 0.70"],
        },
        capabilities: ["core-ios"],
      },
    });

    const result = checkPackageManifest("package.json", defaultOptions);

    expect(result).toBe("success");
    expect(consoleLogSpy).not.toBeCalled();
    expect(consoleWarnSpy).not.toBeCalled();
  });

  test("handles development version ranges", () => {
    fs.__setMockContent({
      ...mockManifest,
      peerDependencies: {
        react: react_v68_v69_v70,
        "react-native": v68_v69_v70,
      },
      devDependencies: {
        react: packageVersion(profile_0_69, "react"),
        "react-native": packageVersion(profile_0_69, "core"),
      },
    });
    rnxKitConfig.__setMockConfig({
      alignDeps: {
        requirements: {
          development: ["react-native@0.69"],
          production: ["react-native@0.68 || 0.69 || 0.70"],
        },
        capabilities: ["core-ios"],
      },
    });

    const result = checkPackageManifest("package.json", defaultOptions);

    expect(result).toBe("success");
    expect(consoleLogSpy).not.toBeCalled();
    expect(consoleWarnSpy).not.toBeCalled();
  });
});

describe("checkPackageManifest({ kitType: 'library' }) (backwards compatibility)", () => {
  const rnxKitConfig = require("@rnx-kit/config");
  const fs = require("fs");

  const mockManifest = {
    name: "@rnx-kit/align-deps",
    version: "0.0.1",
  };

  const react_v68_v69_v70 = [
    packageVersion(profile_0_68, "react"),
    packageVersion(profile_0_69, "react"),
    packageVersion(profile_0_70, "react"),
  ].join(" || ");

  const v68_v69_v70 = [
    packageVersion(profile_0_68, "core"),
    packageVersion(profile_0_69, "core"),
    packageVersion(profile_0_70, "core"),
  ].join(" || ");

  beforeEach(() => {
    fs.__setMockContent({});
    rnxKitConfig.__setMockConfig();
  });

  afterAll(() => {
    jest.clearAllMocks();
  });

  test("returns error code when reading invalid manifests", () => {
    const result = checkPackageManifest("package.json", defaultOptions);
    expect(result).not.toBe("success");
  });

  test("returns early if 'rnx-kit' is missing from the manifest", () => {
    fs.__setMockContent({
      ...mockManifest,
      dependencies: { "react-native-linear-gradient": "0.0.0" },
    });

    const result = checkPackageManifest("package.json", defaultOptions);

    expect(result).toBe("not-configured");
  });

  test("prints warnings when detecting bad packages", () => {
    fs.__setMockContent({
      ...mockManifest,
      dependencies: { "react-native-linear-gradient": "0.0.0" },
      peerDependencies: {
        "react-native": profile_0_70["core"],
      },
      devDependencies: {
        "react-native": profile_0_70["core"],
      },
    });
    rnxKitConfig.__setMockConfig({ reactNativeVersion: "0.70.0" });

    const result = checkPackageManifest("package.json", defaultOptions);

    expect(result).toBe("success");
  });

  test("prints warnings when detecting bad packages (with version range)", () => {
    fs.__setMockContent({
      ...mockManifest,
      dependencies: { "react-native-linear-gradient": "0.0.0" },
    });
    rnxKitConfig.__setMockConfig({ reactNativeVersion: "^0.69.0 || ^0.70.0" });

    const result = checkPackageManifest("package.json", defaultOptions);

    expect(result).toBe("success");
  });

  test("returns early if no capabilities are defined", () => {
    fs.__setMockContent(mockManifest);
    rnxKitConfig.__setMockConfig({ reactNativeVersion: "0.70.0" });

    const result = checkPackageManifest("package.json", defaultOptions);

    expect(result).toBe("success");
  });

  test("returns if no changes are needed", () => {
    fs.__setMockContent({
      ...mockManifest,
      peerDependencies: {
        react: packageVersion(profile_0_70, "react"),
        "react-native": packageVersion(profile_0_70, "core"),
      },
      devDependencies: {
        react: packageVersion(profile_0_70, "react"),
        "react-native": packageVersion(profile_0_70, "core"),
      },
    });
    rnxKitConfig.__setMockConfig({
      reactNativeVersion: "0.70.0",
      capabilities: ["core-ios"],
    });

    const result = checkPackageManifest("package.json", defaultOptions);

    expect(result).toBe("success");
  });

  test("returns if no changes are needed (write: true)", () => {
    let didWriteToPath = false;

    fs.__setMockContent({
      ...mockManifest,
      peerDependencies: {
        react: packageVersion(profile_0_70, "react"),
        "react-native": packageVersion(profile_0_70, "core"),
      },
      devDependencies: {
        react: packageVersion(profile_0_70, "react"),
        "react-native": packageVersion(profile_0_70, "core"),
      },
    });
    fs.__setMockFileWriter((p, _content) => {
      didWriteToPath = p;
    });
    rnxKitConfig.__setMockConfig({
      reactNativeVersion: "0.70.0",
      capabilities: ["core-ios"],
    });

    expect(checkPackageManifest("package.json", writeOptions)).toBe("success");
    expect(didWriteToPath).toBe(false);
  });

  test("returns error code if changes are needed", () => {
    fs.__setMockContent(mockManifest);
    rnxKitConfig.__setMockConfig({
      reactNativeVersion: "0.70.0",
      capabilities: ["core-ios"],
    });

    const result = checkPackageManifest("package.json", defaultOptions);

    expect(result).toBe("unsatisfied");
  });

  test("writes changes back to 'package.json'", () => {
    let didWriteToPath = false;

    fs.__setMockContent(mockManifest);
    fs.__setMockFileWriter((p, _content) => {
      didWriteToPath = p;
    });
    rnxKitConfig.__setMockConfig({
      reactNativeVersion: "0.70.0",
      capabilities: ["core-ios"],
    });

    expect(checkPackageManifest("package.json", writeOptions)).toBe("success");
    expect(didWriteToPath).toBe("package.json");
  });

  test("preserves indentation in 'package.json'", () => {
    let output = "";

    fs.__setMockContent(mockManifest, "\t");
    fs.__setMockFileWriter((_, content) => {
      output = content;
    });
    rnxKitConfig.__setMockConfig({
      reactNativeVersion: "0.70.0",
      capabilities: ["core-ios"],
    });

    expect(checkPackageManifest("package.json", writeOptions)).toBe("success");
    expect(output).toMatchSnapshot();
  });

  test("uses minimum supported version as development version", () => {
    fs.__setMockContent({
      ...mockManifest,
      peerDependencies: {
        react: react_v68_v69_v70,
        "react-native": v68_v69_v70,
      },
      devDependencies: {
        react: packageVersion(profile_0_68, "react"),
        "react-native": packageVersion(profile_0_68, "core"),
      },
    });
    rnxKitConfig.__setMockConfig({
      reactNativeVersion: "^0.68 || ^0.69 || ^0.70",
      capabilities: ["core-ios"],
    });

    const result = checkPackageManifest("package.json", defaultOptions);

    expect(result).toBe("success");
  });

  test("uses declared development version", () => {
    fs.__setMockContent({
      ...mockManifest,
      peerDependencies: {
        react: react_v68_v69_v70,
        "react-native": v68_v69_v70,
      },
      devDependencies: {
        react: packageVersion(profile_0_69, "react"),
        "react-native": packageVersion(profile_0_69, "core"),
      },
    });
    rnxKitConfig.__setMockConfig({
      reactNativeVersion: "^0.68 || ^0.69 || ^0.70",
      reactNativeDevVersion: "0.69.4",
      capabilities: ["core-ios"],
    });

    const result = checkPackageManifest("package.json", defaultOptions);

    expect(result).toBe("success");
  });

  test("handles development version ranges", () => {
    fs.__setMockContent({
      ...mockManifest,
      peerDependencies: {
        react: react_v68_v69_v70,
        "react-native": v68_v69_v70,
      },
      devDependencies: {
        react: packageVersion(profile_0_69, "react"),
        "react-native": packageVersion(profile_0_69, "core"),
      },
    });
    rnxKitConfig.__setMockConfig({
      reactNativeVersion: "^0.68 || ^0.69 || ^0.70",
      reactNativeDevVersion: "^0.69.4",
      capabilities: ["core-ios"],
    });

    const result = checkPackageManifest("package.json", defaultOptions);

    expect(result).toBe("success");
  });
});
