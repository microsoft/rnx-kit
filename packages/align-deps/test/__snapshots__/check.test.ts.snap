// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checkPackageManifest({ kitType: 'library' }) (backwards compatibility) preserves indentation in 'package.json' 1`] = `
"{
	\\"name\\": \\"@rnx-kit/align-deps\\",
	\\"version\\": \\"0.0.1\\",
	\\"peerDependencies\\": {
		\\"react\\": \\"18.1.0\\",
		\\"react-native\\": \\"^0.70.0\\"
	},
	\\"devDependencies\\": {
		\\"react\\": \\"18.1.0\\",
		\\"react-native\\": \\"^0.70.0\\"
	}
}
"
`;

exports[`checkPackageManifest({ kitType: 'library' }) preserves indentation in 'package.json' 1`] = `
"{
	\\"name\\": \\"@rnx-kit/align-deps\\",
	\\"version\\": \\"0.0.1\\",
	\\"peerDependencies\\": {
		\\"react\\": \\"18.1.0\\",
		\\"react-native\\": \\"^0.70.0\\"
	},
	\\"devDependencies\\": {
		\\"react\\": \\"18.1.0\\",
		\\"react-native\\": \\"^0.70.0\\"
	}
}
"
`;
